Control digitizer mouse single interface

Skyler84 opened the pull request:
Added to James' work on combining the HID reports into a single interface along with the control interface.
All changes from AXPB009 have been merged into this branch so it is up to date.

Jimmy-Cameron committed:
Made a single interface USB device (TBP Control interface).

Jimmy-Cameron committed:
Control interface functioning correctly with report ID.

Sacrifices a byte per transfer.

Jimmy-Cameron committed:
Digitizer moved over to control endpoint with a different report ID.

Jimmy-Cameron committed:
Updated revision number to reflect engineering build.

Commit:
Merge branch 'main' into control-digitizer-single-interface

Commit:
Add support for combined mouse HID

AbsMouse HID interface can now be used with the single combined Command/Digitizer interface with mode switching.

Commit:
Merge branch 'main' into control-digitizer-mouse-single-interface

Jimmy-Cameron commented on Middlewares/ST/STM32_USB_Device_Library/Class/HID/HID_Mouse/usbd_mouse.c:
```diff
@@ -177,16 +177,18 @@ void GetMouseDescriptorLength(uint8_t BridgeMode)
-    }
+    NumInterfaces = 1;
+
+//    if(BridgeMode == ABSOLUTE_MOUSE || BridgeMode == PARALLEL_DIGITIZER)  // mouse/digitizer enabled
```
I know I was the one to do it, but can you remove this unused code please.

Jimmy-Cameron commented on Src/Flash_Control.c:
```diff
@@ -239,31 +239,32 @@ void Store_BridgeMode_To_Flash(uint8_t BridgeMode_to_store)
-        Store_BridgeMode_To_Flash(MODE_PARALLEL_DIGITIZER);
-    }
-
+   uint8_t BridgeMode_temp    = 0;
```
By the looks of things these need to be indented by another space.

Jimmy-Cameron commented on Src/_AXPB-009_Main.c:
```diff
@@ -200,7 +200,7 @@ int main(void)
         }
 
-        if((boMouseReportToSend == 1) && (USBD_MOUSE_HID_GetState(&hUsbDeviceFS) == USB_HID_IDLE) && (usb_remote_wake_state == RESUMED))
+        if((boMouseReportToSend == 1) && (USBD_GENERIC_HID_GetState(&hUsbDeviceFS) == USB_HID_IDLE) && (usb_remote_wake_state == RESUMED))
```
I'm unsure as to the purpose of doing so, but there could potentially be an issue here when the bridge has been configured for proxy mode (reading all reports from aXiom and sending them to the host via report id 1) and is also sending digitizer reports (report id 2). There may be a case where the digitizer never has a chance to send a report, as it comes after the control reports. A way around this would be to use the circular buffer for all reports.

Jimmy-Cameron commented on Src/_AXPB-009_Main.c:
I recall someone mentioning that the 3D raw data view wasn't working with these report IDs. Has that been addressed? Feels like it could either be fixed here or in TH2.

Skyler84 replied:
3D raw data does appear to be broken. I'm not sure exactly what causes it currently.

Commit:
remove unused code

Commit:
Limit control interface to 5 transmit at a time.
Allows mouse/digitizer interface to send its report.

Commit:
fix indentation

Commit:
Fix 3d raw data

HID report ID adds one byte causing last byte to be cut off.
Reduced max RX bytes by one to accomodate.

Skyler84 commented:
Reduced the RX byte count by one fixes 3D Raw data (works in touchHub!)

Jimmy-Cameron commented on Inc/USB/usbd_conf.h:
```diff
@@ -62,14 +62,21 @@ extern PCD_HandleTypeDef hpcd_USB_FS;
-#define USBD_GENERIC_HID_REPORT_IN_SIZE         (64)// limited by USB 2.0 spec, can't be any bigger!
+#define PRESS_INTERFACE_ENABLED 1
+#define MOUSE_INTERFACE_ENABLED 1
+#define DIGITIZER_HID_ENABLED 1
```
DIGITIZER_HID_ENABLED and MOUSE_HID_ENABLED aren't used anywhere.

Jimmy-Cameron commented on Middlewares/ST/STM32_USB_Device_Library/Class/Composite/usbd_composite.c:
```diff
@@ -147,8 +147,7 @@ __ALIGN_BEGIN uint8_t USBD_COMPOSITE_HID_CfgDesc[USB_COMPOSITE_HID_CONFIG_DESC_S
         0x22,                               /* bDescriptorType*/
-        USBD_GENERIC_HID_REPORT_DESC_SIZE,  /* wItemLength: Total length of Report descriptor*/
-        0x00,
+        LOBYTE(USBD_GENERIC_HID_REPORT_DESC_SIZE), HIBYTE(USBD_GENERIC_HID_REPORT_DESC_SIZE), /* wItemLength: Total length of Report descriptor*/
```
I've just realised the full configuration descriptor is here. Since we only need a single interface, you can remove the unused interfaces for tidiness.

Jimmy-Cameron commented on Middlewares/ST/STM32_USB_Device_Library/Class/HID/HID_Mouse/usbd_mouse.c:
```diff
@@ -177,16 +177,7 @@ void GetMouseDescriptorLength(uint8_t BridgeMode)
-        boMouseEnabled = false;
-        NumInterfaces = 2;
-    }
+    NumInterfaces = 1;
```
Just documenting that this function no longer serves a purpose, but it's not an issue to leave as it.

Jimmy-Cameron commented on Src/Flash_Control.c:
```diff
@@ -264,6 +264,7 @@ uint8_t GetDeviceModeFromFlash(void)
         Store_BridgeMode_To_Flash(MODE_PARALLEL_DIGITIZER);
     }
 
+    // return MODE_PARALLEL_DIGITIZER;
```
Delete commented code.

Jimmy-Cameron commented on Inc/USB/usbd_conf.h:
```diff
@@ -62,14 +62,21 @@ extern PCD_HandleTypeDef hpcd_USB_FS;
-/*---------- -----------*/
-#define USBD_GENERIC_HID_REPORT_IN_SIZE         (64)// limited by USB 2.0 spec, can't be any bigger!
+#define PRESS_INTERFACE_ENABLED 1
+#define MOUSE_INTERFACE_ENABLED 1
```
Not strictly necessary as this device will always only be a single interface, but please commit these as disabled.

Jimmy-Cameron reviewed (changes requested):
One general comment, whilst closely related to the axpb009 this project should probably be called axpb015. Please go ahead and rename it, also making sure it appears as AXPB015 to the host. This will break TH2, but talk to Tim to get it added to the list of supported bridges.

Jimmy-Cameron commented:
Once renamed and the final few bits addressed (either code change or just ignore, up to you) I am happy for this to merge. Tested and everything works as expected.

Commit:
remodelling: tidy up a few things left behind.

Commit:
Rename USB descriptors to AXPB015

Skyler84 commented:
Should all be done!
